connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala

```diff
@@ -829,10 +829,23 @@ object SparkSession extends Logging {
 
   /**
    * Set the (global) default [[SparkSession]], and (thread-local) active [[SparkSession]] when
-   * they are not set yet.
+   * they are not set yet or the associated [[SparkConnectClient]] is unusable.
    */
   private def setDefaultAndActiveSession(session: SparkSession): Unit = {
-    defaultSession.compareAndSet(null, session)
+    var currentDefault = defaultSession.getAcquire
+    var swapped = false
+    while ((currentDefault == null ||
+        (currentDefault.client != null && currentDefault.client.hasSessionChanged))
```

Comment:
null check is not necessary. safe to assume that every session will have a client attached.

Reply:
Ack, I'll fix it!